File: js/scorio/src/rank/evalRanking.ts
/**
 * Evaluation-metric ranking methods. Port of `scorio/rank/eval_ranking.py`.
 *
 * Each method maps a model's responses to a scalar score using the ported
 * `scorio/eval` metric of the same name, then converts scores to ranks.
 */

import {
  avg as evalAvg,
  bayes as evalBayes,
  passAtK as evalPassAtK,
  passHatK as evalPassHatK,
  gPassAtKTau as evalGPassAtKTau,
  mgPassAtK as evalMgPassAtK,
} from "../eval/index.js";
import { rankScores } from "./internal/rankScores.js";
import { normPpf } from "./internal/special.js";
import {
  validateInput,
  shape3,
  type TensorInput,
} from "./internal/tensor.js";
import type { BaseRankOptions, RankResult } from "./internal/result.js";

const defaultIfUndefined = <T>(value: T | undefined, fallback: T): T =>
  value === undefined ? fallback : value;

function coercePythonFloat(value: unknown, name: string): number {
  if (typeof value === "number") return value;
  if (typeof value === "boolean") return value ? 1 : 0;
  if (typeof value === "string" && value.trim().length > 0) {
    const normalized = value.trim().replace(/(?<=\d)_(?=\d)/g, "");
    if (/^[+-]?(?:inf(?:inity)?)$/i.test(normalized)) {
      return normalized.startsWith("-") ? -Infinity : Infinity;
    }
    if (/^[+-]?nan$/i.test(normalized)) return NaN;
    if (/^[+-]?(?:(?:\d+(?:\.\d*)?|\.\d+)(?:e[+-]?\d+)?)$/i.test(normalized)) {
      return Number(normalized);
    }
  }
  throw new TypeError(`${name} must be convertible to a float`);
}

/** Python's comparison-based pass-family checks accept numbers and booleans. */
function comparableNumber(value: unknown, name: string): number {
  if (typeof value === "boolean") return value ? 1 : 0;
  if (typeof value === "number") return value;
  throw new TypeError(`${name} must be a real scalar`);
}

/** Rank models by mean accuracy over all questions and trials (`eval.avg`). */
export function avg(R: TensorInput, options: BaseRankOptions = {}): RankResult {
  const method = defaultIfUndefined(options.method, "competition");
  const tensor = validateInput(R);
  const [L] = shape3(tensor);
  const scores = Array.from({ length: L }, (_, l) => evalAvg(tensor[l]!)[0]);
  return { ranking: rankScores(scores, method), scores };
}

/** Options for {@link bayes}. */
export interface BayesRankOptions extends BaseRankOptions {
  /** Category-to-score weight vector of length `C+1`. Required for non-binary `R`. */
  w?: readonly number[] | null;
  /** Prior outcomes: shared `(M, D)` or per-model `(L, M, D)`. */
  R0?: TensorInput | null;
  /** Posterior quantile `q` in `(0, 1)`; if set, rank by `mu + Φ⁻¹(q)·sigma`. */
  quantile?: number | null;
}

function priorOutcome(value: unknown): number {
  if (typeof value === "boolean") return value ? 1 : 0;
  if (typeof value !== "number" || !Number.isFinite(value) || !Number.isInteger(value)) {
    throw new Error("R0 must contain real, finite integer-valued outcomes");
  }
  return value;
}

/** Rank models with Bayes@N posterior statistics (`eval.bayes`). */
export function bayes(R: TensorInput, options: BayesRankOptions = {}): RankResult {
  const method = defaultIfUndefined(options.method, "competition");
  const tensor = validateInput(R, false);
  const [L, M] = shape3(tensor);

  let quantile: number | undefined;
  if (options.quantile !== undefined && options.quantile !== null) {
    quantile = coercePythonFloat(options.quantile, "quantile");
    if (!Number.isFinite(quantile) || !(quantile > 0 && quantile < 1)) {
      throw new Error(`quantile must be in (0, 1); got ${options.quantile}`);
    }
  }

  let r0Shared: number[][] | undefined;
  let r0PerModel: number[][][] | undefined;
  if (options.R0 !== undefined && options.R0 !== null) {
    const raw = options.R0 as unknown;
    if (!Array.isArray(raw)) {
      throw new Error("R0 must be shape (M, D) or (L, M, D)");
    }
    const is2d = raw.every(
      (row) =>
        Array.isArray(row) &&
        row.every((value) => !Array.isArray(value)),
    );
    const is3d = raw.every(
      (matrix) =>
        Array.isArray(matrix) &&
        matrix.every(
          (row) =>
            Array.isArray(row) &&
            row.every((value) => !Array.isArray(value)),
        ),
    );
    if (is2d) {
      const mat = raw.map((row) => (row as unknown[]).map(priorOutcome));
      const D = mat[0]?.length ?? 0;
      if (mat.some((row) => row.length !== D)) {
        throw new Error("R0 must be a rectangular array");
      }
      if (mat.length !== M) {
        throw new Error(`Shared R0 must have shape (M=${M}, D)`);
      }
      r0Shared = mat;
    } else if (is3d) {
      const t = raw.map((matrix) =>
        (matrix as unknown[][]).map((row) => row.map(priorOutcome)),
      );
      const d = t[0]?.[0]?.length ?? 0;
      if (
        t.some(
          (matrix) =>
            matrix.length !== (t[0]?.length ?? 0) ||
            matrix.some((row) => row.length !== d),
        )
      ) {
        throw new Error("R0 must be a rectangular array");
      }
      if (t.length !== L || (t[0]?.length ?? 0) !== M) {
        throw new Error(`Model-specific R0 must have shape (L=${L}, M=${M}, D)`);
      }
      r0PerModel = t;
    } else {
      throw new Error("R0 must be shape (M, D) or (L, M, D)");
    }
  }

  const z = quantile !== undefined ? normPpf(quantile) : undefined;
  const scores = new Array<number>(L).fill(0);
  for (let l = 0; l < L; l++) {
    const modelR0 = r0PerModel ? r0PerModel[l]! : r0Shared;
    const [mu, sigma] = evalBayes(tensor[l]!, options.w, modelR0);
    scores[l] = z !== undefined ? mu + z * sigma : mu;
  }
  return { ranking: rankScores(scores, method), scores };
}

/** Rank models by Pass@k (`eval.pass_at_k`). */
export function passAtK(
  R: TensorInput,
  k: number,
  options: BaseRankOptions = {},
): RankResult {
  const method = defaultIfUndefined(options.method, "competition");
  const tensor = validateInput(R);
  const normalizedK = comparableNumber(k, "k");
  const [L] = shape3(tensor);
  const scores = Array.from({ length: L }, (_, l) =>
    evalPassAtK(tensor[l]!, normalizedK),
  );
  return { ranking: rankScores(scores, method), scores };
}

/** Rank models by Pass-hat@k / G-Pass@k (`eval.pass_hat_k`). */
export function passHatK(
  R: TensorInput,
  k: number,
  options: BaseRankOptions = {},
): RankResult {
  const method = defaultIfUndefined(options.method, "competition");
  const tensor = validateInput(R);
  const normalizedK = comparableNumber(k, "k");
  const [L] = shape3(tensor);
  const scores = Array.from({ length: L }, (_, l) =>
    evalPassHatK(tensor[l]!, normalizedK),
  );
  return { ranking: rankScores(scores, method), scores };
}

/** Rank models by generalized G-Pass@k_tau (`eval.g_pass_at_k_tau`). */
export function gPassAtKTau(
  R: TensorInput,
  k: number,
  tau: number,
  options: BaseRankOptions = {},
): RankResult {
  const method = defaultIfUndefined(options.method, "competition");
  const tensor = validateInput(R);
  const normalizedTau = comparableNumber(tau, "tau");
  const normalizedK = comparableNumber(k, "k");
  const [L] = shape3(tensor);
  const scores = Array.from({ length: L }, (_, l) =>
    evalGPassAtKTau(tensor[l]!, normalizedK, normalizedTau),
  );
  return { ranking: rankScores(scores, method), scores };
}

/** Rank models by mG-Pass@k (`eval.mg_pass_at_k`). */
export function mgPassAtK(
  R: TensorInput,
  k: number,
  options: BaseRankOptions = {},
): RankResult {
  const method = defaultIfUndefined(options.method, "competition");
  const tensor = validateInput(R);
  const normalizedK = comparableNumber(k, "k");
  const [L] = shape3(tensor);
  const scores = Array.from({ length: L }, (_, l) =>
    evalMgPassAtK(tensor[l]!, normalizedK),
  );
  return { ranking: rankScores(scores, method), scores };
}

File: julia/Scorio.jl/src/rank/eval_ranking.jl
"""Eval-metric-based ranking methods."""

using SpecialFunctions: erfcinv

"""
    avg(R; method="competition", return_scores=false)

Rank models by per-model mean accuracy across all questions and trials.

For each model `l`, compute the scalar score:

```math
s_l^{\\mathrm{avg}} = \\frac{1}{MN}\\sum_{m=1}^{M}\\sum_{n=1}^{N} R_{lmn}
```

Higher scores are better; ranking is produced by `rank_scores`.

# Arguments
- `R`: binary response tensor `(L, M, N)` or matrix `(L, M)` promoted to `(L, M, 1)`.
- `method`: tie-handling rule for `rank_scores`.
- `return_scores`: if `true`, return `(ranking, scores)`.
"""
function _rank_avg(R; method="competition", return_scores=false)
    Rv = validate_input(R)
    L = size(Rv, 1)

    scores = zeros(Float64, L)
    for model in 1:L
        mu, _ = avg(@view Rv[model, :, :])
        scores[model] = mu
    end

    ranking = rank_scores(scores)[string(method)]
    return return_scores ? (ranking, scores) : ranking
end

avg(R::AbstractArray{<:Any,3}; method="competition", return_scores=false) =
    _rank_avg(R; method=method, return_scores=return_scores)

function _norm_ppf(p::Float64)::Float64
    p == 0.0 && return -Inf
    p == 1.0 && return Inf
    return -sqrt(2.0) * erfcinv(2.0 * p)
end

"""
    bayes(
        R::AbstractArray{<:Integer, 3},
        w=nothing;
        R0=nothing,
        quantile=nothing,
        method="competition",
        return_scores=false,
    )

Rank models by Bayes@N scores computed independently per model.

If `quantile` is provided, models are ranked by `mu + z_q * sigma`; otherwise
by posterior mean `mu`.

# References
Hariri, M., Samandar, A., Hinczewski, M., & Chaudhary, V. (2026).
Don't Pass@k: A Bayesian Framework for Large Language Model Evaluation.
*The Fourteenth International Conference on Learning Representations*.
https://openreview.net/forum?id=PTXi3Ef4sT

# Formula
For each model `l`, let `(mu_l, sigma_l) = Scorio.bayes(R_l, w, R0_l)`.

```math
s_l =
\\begin{cases}
\\mu_l, & \\text{if quantile is not set} \\\\
\\mu_l + \\Phi^{-1}(q)\\,\\sigma_l, & \\text{if quantile}=q \\in [0,1]
\\end{cases}
```

# Arguments
- `R`: integer tensor `(L, M, N)` with values in `{0, ..., C}`.
- `w`: class weights of length `C+1`. If not provided and R is binary (contains only 0 and 1),
  defaults to `[0.0, 1.0]`. For non-binary R, w is required.
- `R0`: optional shared prior `(M, D)` or model-specific prior `(L, M, D)`.
- `quantile`: optional value in `[0, 1]` for quantile-adjusted ranking.
- `method`, `return_scores`: ranking output controls.
"""
function _rank_bayes(
    R,
    w=nothing;
    R0=nothing,
    quantile=nothing,
    method="competition",
    return_scores=false,
)
    Rv = validate_input(R; binary_only=false)
    L, M, _ = size(Rv)

    z = nothing
    if !isnothing(quantile)
        q = Float64(quantile)
        if !isfinite(q) || !(0.0 < q < 1.0)
            error("quantile must be in (0, 1); got $quantile")
        end
        z = _norm_ppf(q)
    end

    R0_shared = nothing
    R0_per_model = nothing

    if !isnothing(R0)
        raw_R0 = _coerce_rank_array_like(R0)
        if isnothing(raw_R0)
            error("R0 must contain real, finite integer-valued outcomes")
        end
        if !(eltype(raw_R0) <: Number) || eltype(raw_R0) <: Complex ||
           any(x -> !isfinite(x) || x != floor(x), raw_R0)
            error("R0 must contain real, finite integer-valued outcomes")
        end
        R0_arr = Int.(raw_R0)

        if ndims(R0_arr) == 2
            if size(R0_arr, 1) != M
                error("Shared R0 must have shape (M=$M, D), got $(size(R0_arr))")
            end
            R0_shared = R0_arr
        elseif ndims(R0_arr) == 3
            if size(R0_arr, 1) != L || size(R0_arr, 2) != M
                error(
                    "Model-specific R0 must have shape (L=$L, M=$M, D), got $(size(R0_arr))",
                )
            end
            R0_per_model = R0_arr
        else
            error(
                "R0 must be shape (M, D) or (L, M, D); got ndim=$(ndims(R0_arr)) with shape $(size(R0_arr))",
            )
        end
    end

    scores = zeros(Float64, L)
    for model in 1:L
        model_R0 = isnothing(R0_shared) ? nothing : R0_shared
        if !isnothing(R0_per_model)
            model_R0 = @view R0_per_model[model, :, :]
        end

        mu, sigma = bayes(@view(Rv[model, :, :]), w, model_R0)
        scores[model] = isnothing(z) ? mu : (mu + z * sigma)
    end

    ranking = rank_scores(scores)[string(method)]
    return return_scores ? (ranking, scores) : ranking
end

function bayes(
    R::AbstractArray{<:Integer,3},
    w=nothing;
    R0=nothing,
    quantile=nothing,
    method="competition",
    return_scores=false,
)
    return _rank_bayes(
        R,
        w;
        R0=R0,
        quantile=quantile,
        method=method,
        return_scores=return_scores,
    )
end

"""
    pass_at_k(R::AbstractArray{<:Integer, 3}, k; method="competition", return_scores=false)

Rank models by per-model Pass@k scores.

For each model `l`, define per-question success counts
``nu_{lm} = \\sum_{n=1}^{N} R_{lmn}``. Then:

```math
s_l^{\\mathrm{Pass@}k}
= \\frac{1}{M}\\sum_{m=1}^{M}
\\left(1 - \\frac{\\binom{N-\\nu_{lm}}{k}}{\\binom{N}{k}}\\right)
```

# References
Chen, M., Tworek, J., Jun, H., et al. (2021).
Evaluating Large Language Models Trained on Code.
*arXiv:2107.03374*. https://arxiv.org/abs/2107.03374
"""
function _rank_pass_at_k(
    R,
    k;
    method="competition",
    return_scores=false,
)
    Rv = validate_input(R)
    L = size(Rv, 1)

    scores = zeros(Float64, L)
    for model in 1:L
        scores[model] = pass_at_k(@view(Rv[model, :, :]), k)
    end

    ranking = rank_scores(scores)[string(method)]
    return return_scores ? (ranking, scores) : ranking
end

pass_at_k(R::AbstractArray{<:Integer,3}, k; method="competition", return_scores=false) =
    _rank_pass_at_k(R, k; method=method, return_scores=return_scores)

"""
    pass_hat_k(R::AbstractArray{<:Integer, 3}, k; method="competition", return_scores=false)

Rank models by per-model Pass-hat@k (G-Pass@k) scores.

With ``nu_{lm} = \\sum_{n=1}^{N} R_{lmn}``:

```math
s_l^{\\widehat{\\mathrm{Pass@}k}}
= \\frac{1}{M}\\sum_{m=1}^{M}
\\frac{\\binom{\\nu_{lm}}{k}}{\\binom{N}{k}}
```

# References
Yao, S., Shinn, N., Razavi, P., & Narasimhan, K. (2024).
tau-bench: A Benchmark for Tool-Agent-User Interaction in Real-World Domains.
*arXiv:2406.12045*. https://arxiv.org/abs/2406.12045
"""
function _rank_pass_hat_k(
    R,
    k;
    method="competition",
    return_scores=false,
)
    Rv = validate_input(R)
    L = size(Rv, 1)

    scores = zeros(Float64, L)
    for model in 1:L
        scores[model] = pass_hat_k(@view(Rv[model, :, :]), k)
    end

    ranking = rank_scores(scores)[string(method)]
    return return_scores ? (ranking, scores) : ranking
end

pass_hat_k(R::AbstractArray{<:Integer,3}, k; method="competition", return_scores=false) =
    _rank_pass_hat_k(R, k; method=method, return_scores=return_scores)

"""
    g_pass_at_k_tau(
        R::AbstractArray{<:Integer, 3},
        k,
        tau;
        method="competition",
        return_scores=false,
)

Rank models by generalized G-Pass@k_τ per model.

Let ``X_{lm} ~ Hypergeometric(N, nu_{lm}, k)`` where
``nu_{lm} = \\sum_{n=1}^{N} R_{lmn}``. The score is:

```math
s_l^{\\mathrm{G\\text{-}Pass@}k_{\\tau}}
= \\frac{1}{M}\\sum_{m=1}^{M}
\\Pr\\!\\left(X_{lm}\\ge \\lceil \\tau k \\rceil\\right)
```

```math
\\Pr(X_{lm}\\ge \\lceil \\tau k \\rceil)
= \\sum_{j=\\lceil \\tau k \\rceil}^{k}
\\frac{\\binom{\\nu_{lm}}{j}\\binom{N-\\nu_{lm}}{k-j}}{\\binom{N}{k}}
```

# References
Liu, J., Liu, H., Xiao, L., et al. (2024).
Are Your LLMs Capable of Stable Reasoning?
*arXiv:2412.13147*. https://arxiv.org/abs/2412.13147
"""
function _rank_g_pass_at_k_tau(
    R,
    k,
    tau;
    method="competition",
    return_scores=false,
)
    Rv = validate_input(R)
    L = size(Rv, 1)

    scores = zeros(Float64, L)
    for model in 1:L
        scores[model] = g_pass_at_k_tau(@view(Rv[model, :, :]), k, tau)
    end

    ranking = rank_scores(scores)[string(method)]
    return return_scores ? (ranking, scores) : ranking
end

function g_pass_at_k_tau(
    R::AbstractArray{<:Integer,3},
    k,
    tau;
    method="competition",
    return_scores=false,
)
    return _rank_g_pass_at_k_tau(R, k, tau; method=method, return_scores=return_scores)
end

"""
    mg_pass_at_k(R::AbstractArray{<:Integer, 3}, k; method="competition", return_scores=false)

Rank models by per-model mG-Pass@k scores.

With ``X_{lm} ~ Hypergeometric(N, nu_{lm}, k)`` and ``m_0 = \\lceil k/2 \\rceil``:

```math
s_l^{\\mathrm{mG\\text{-}Pass@}k}
= \\frac{1}{M}\\sum_{m=1}^{M}
\\frac{2}{k}\\,\\mathbb{E}\\!\\left[(X_{lm}-m_0)_+\\right]
```

Equivalent discrete form:

```math
\\frac{2}{k}\\sum_{i=m_0+1}^{k}\\Pr(X_{lm}\\ge i)
```

# References
Liu, J., Liu, H., Xiao, L., et al. (2024).
Are Your LLMs Capable of Stable Reasoning?
*arXiv:2412.13147*. https://arxiv.org/abs/2412.13147
"""
function _rank_mg_pass_at_k(
    R,
    k;
    method="competition",
    return_scores=false,
)
    Rv = validate_input(R)
    L = size(Rv, 1)

    scores = zeros(Float64, L)
    for model in 1:L
        scores[model] = mg_pass_at_k(@view(Rv[model, :, :]), k)
    end

    ranking = rank_scores(scores)[string(method)]
    return return_scores ? (ranking, scores) : ranking
end

mg_pass_at_k(R::AbstractArray{<:Integer,3}, k; method="competition", return_scores=false) =
    _rank_mg_pass_at_k(R, k; method=method, return_scores=return_scores)
